---
title: "BMIN503/EPID600 Project Template"
author: "Your Name"
output: 
  html_document:
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
smallpox <- read.table("https://raw.githubusercontent.com/roopshab/BMIN503_Final_Project/master/GSE183621_Fold_change.txt")
```  
***
Use this template to complete your project throughout the course. Your Final Project presentation will be based on the contents of this document. Replace the title/name above and text below with your own, but keep the headers.

Name: Roopsha Bandopadhyay 

### Overview
Give a brief a description of your project and its goal(s), what data you are using to complete it, and what three faculty/staff in different fields you have spoken to about your project with a brief summary of what you learned from each person. Include a link to your final project GitHub repository.

```{r}
# Link to repository: https://github.com/roopshab/BMIN503_Final_Project
# The goal of this project is to determine if the percentage of vaccinated people required for effective herd-immunity against certain contagious diseases varies over geographical areas of different sizes. This project will also evaluate factors that cause this variation in percentage.
# Data used: RamiKrispin
# I spoke with Prof. Gary Smith and learned about epidemiological principles of vector spread.
```

### Introduction 
Describe the problem addressed, its significance, and some background to motivate the problem. Explain why your problem is interdisciplinary, what fields can contribute to its understanding, and incorporate background related to what you learned from meeting with faculty/staff.

```{r}

```

### Methods
Describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why. 

### Biology-based factors ###

### COVID Variant Data
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

for (pkg in c("GEOquery", "oligo", "limma", "hgu133plus2.db", "pd.hg.u133.plus.2", "viridis", "fgsea")) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    BiocManager::install(pkg)
  }
}

for (pkg in c("gplots")) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    install.packages(pkg)
  }
}
```

* Use GSEA to check which functional pathways are involved in smoking response. Use KEGG and REACTOME pathways from MSigDB as reference ontologies. Create barplots of NES for **independent** pathways with adjusted p-value < 0.05 and absolute NES > 2. (*3 points*)
```{r, error=TRUE, warning=FALSE, message=FALSE}
library(fgsea)
library(data.table)
library(ggplot2)

# Import qPCR experiment data from Github and store in data frame.
# Remove all entries with non-significant values (p<0.05)
# Rank the remaining entries by smalles p-value first.
# For the remaining data, plot like this:
# gene  variantname   foldchange  pvalue
# Plot the fold-changes of remaining genes on a bar graph. Color by variant name.
# Read in pathway files in .gmt format using the function fgsea::gmtPathways
kegg <- gmtPathways("./c2.cp.kegg.v7.4.symbols.gmt")
reactome <- gmtPathways("./c2.cp.reactome.v7.4.symbols.gmt")
pathways <- c(kegg, reactome)
####### Read in the differential expression result file
DE_res <- read.table("./GSE27973_DE_results.txt", header = T, sep = "\t")
# Include pathways with >=3 and <=500 genes.
fgseaRes <- fgsea(pathways = pathways, stats = gene_list, minSize = 3, maxSize = 500)
fgseaRes <- fgseaRes[order(fgseaRes$pval), ]


# Rank the gene list based on their t-statistics
gene_list <- DE_res$t
names(gene_list) <- DE_res$SYMBOL
gene_list <- sort(gene_list, decreasing = T)

# Include pathways with >=3 and <=500 genes.
fgseaRes <- fgsea(pathways = pathways, stats = gene_list, minSize = 3, maxSize = 500)
fgseaRes <- fgseaRes[order(fgseaRes$pval), ]

# Select independent pathways using fgsea::collapsePathways
# Obtain final GSEA output
collapsedPathways <- collapsePathways(fgseaRes = fgseaRes[padj < 0.05], pathways = pathways, stats = gene_list)
mainPathways <- fgseaRes[pathway %in% collapsedPathways$mainPathways] # keep results of independent pathways

# Select top-ranked significant pathways with p-value<0.05 and absolute NES>2 
top_pathway <- mainPathways %>%
  filter(padj < 0.05 & abs(NES) > 2) %>%
  arrange(NES)

# Convert pathways to factors where the levels are the same as pathway order
top_pathway$pathway <- factor(top_pathway$pathway, levels = top_pathway$pathway)
ggplot(top_pathway, aes(y = NES, x = pathway)) +
  geom_bar(width = 0.8, position = position_dodge(width = 0.8), stat = "identity", fill = "blue") + 
  coord_flip() +
  theme_bw() +
  theme(axis.title.y = element_blank())
```


### Smallpox Data

* Download the raw CEL files from GEO, create an expression set, and include a phenotype variable corresponding to three levels: control, ectromelia virus at 2 hours, or ectromelia virus at 16 hours.
```{r, error=TRUE, warning=FALSE, message=FALSE}
library(BiocManager)
library(GEOquery)
library(oligo)
library(limma)
library(viridis)

# Download raw CEL files from GEO
if (!file.exists("./GSE120615/GSE120615_RAW.tar"))
  getGEOSuppFiles("GSE120615")

# Create expression set and define phenotype variable
untar("./GSE120615/GSE120615_RAW.tar", exdir = "./GSE120615/data")
celFiles <- list.celfiles("./GSE120615/data", full.names = TRUE, listGzipped = TRUE)
raw.data <- read.celfiles(celFiles)

# Include a phenotype variable corresponding to three levels: control, ectromelia virus at 2 hours, or ectromelia virus at 16 hours.
pData(raw.data)$phenotype <- c(rep(c("Control", "TwoHours", "SixteenHours")))
```

* Look at a boxplot of expression values for each sample and see whether any seem to be outliers.
```{r, error=TRUE, warning=FALSE, message=FALSE}
boxplot(raw.data, col = "red", main = "Raw Probe Intensities")
# The medians of each boxplot are very close to each other; there are no outliers.
```

* Perform RMA on the samples and create a new intensity boxplot.
```{r, error=TRUE, warning=FALSE, message=FALSE}
# Perform RMA on the samples.
GSE120615.rma <- rma(raw.data)

# Create a new intensity boxplot.
boxplot(GSE120615.rma, col = "blue", main = "RMA Expression Values")
```

* Perform a differential expression analysis, and report how many probes are significantly differentially expressed between the control cells and cells with virus at 16 hours based on a Benjamini-Hochberg corrected p-value threshold of 0.05.
```{r, error=TRUE, warning=FALSE, message=FALSE}
library(ggplot2)
library(dplyr)

# Create a model matrix
GSE120615.rma <- rma(raw.data)
design <- model.matrix(~ -1 + factor(GSE120615.rma$phenotype))
colnames(design) <- levels(factor(GSE120615.rma$phenotype))

# Fit a linear model with limma package. Expression data linked to outcome of a design matrix model
fit <- lmFit(GSE120615.rma, design)

# Create contrast groups of interest - control vs. 16H
GSE120615.contrast <- makeContrasts(phenotype = Control - SixteenHours, levels = design)

# Get the contrasts for samples of interest
fit2 <- contrasts.fit(fit, GSE120615.contrast)

# Adjust fit coefficients using an empirical Bayes moderation of standard errors
fit2 <- eBayes(fit2)

# Extract results for each gene by setting `num = Inf` 
treatment_results <- topTable(fit2, coef = "phenotype", adjust = "BH", num = Inf)
test<- treatment_results

# Add gene names
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("mouse4302.db")

library(mouse4302.db)
treatment_results$ID = row.names(treatment_results)
treatment_results$SYMBOL <- sapply(treatment_results$ID, function(x) mget(x, env = mouse4302SYMBOL, ifnotfound = NA)[[1]])

# Save differential expression results of all genes
write.table(treatment_results, "./GSE120615_DE_results.txt", row.names = F, col.names = T, quote = F, sep = "\t")

# Assign (in)significant genes
treatment_results$sig <- rep("insignificant", nrow(treatment_results))
treatment_results$sig[which(treatment_results$P.Value<0.05)] <- "significant"

# Visualize genes using a volcano plot
ggplot(treatment_results, aes(x = logFC, y = -log10(P.Value), color = sig)) + 
  geom_point() +
  theme_bw() +
  ggtitle("Volcano plot") +
  xlab("LogFC")+
  ylab("-Log10(q-value)") +
  scale_color_manual(values = c("black", "red")) +
  theme(legend.position = "none")

```

```{r, error=TRUE, warning=FALSE, message=FALSE}
# Number of significantly differentially expressed probes
significantProbes <- sum(treatment_results$sig == "significant")
significantProbes

# There are 102 significantly differentially expressed probes between the control and virus at 16 hours samples.
```

* What is the top differentially expressed probe and what gene does it correspond to? Create a boxplot of the probe's normalized intensities across **all** groups.
```{r, error=TRUE, warning=FALSE, message=FALSE}
# The top differentially expressed probe is the one with the lowest p-value.
# This is "1430340_at" and corresponds to the Lca5 gene.
  
# Make a data frame for gene expression of the most significant differentially expressed probe "201468_s_at".
df_boxplot = data.frame(
  expression = exprs(GSE120615.rma)["1430340_at", ],
  phenotype = GSE120615.rma$phenotype)

# Boxplot of normalized intensities across all treatment groups of probe 201468_s_at
ggplot(df_boxplot, aes(x = phenotype, y = expression)) +
  geom_boxplot(outlier.colour = NA, color = "grey18", fill = "lightblue") +
  stat_boxplot(geom = "errorbar", color = "grey18") +
  geom_jitter(size = 1, position = position_jitter(width = 0.3)) +
  ggtitle("Expression Values for Probe 1430340_at") +
  xlab(" ") +
  ylab("RMA Intensity") +
  theme_bw() +
  theme(legend.position = "none")
```

* Create a heatmap using the significantly differentially expressed genes obtained above and samples from all conditions. How do the treatment groups cluster? What does this indicate (a qualitative answer is sufficient)?
```{r, error=TRUE, warning=FALSE, message=FALSE}
library(gplots)

# Keep top 102 genes. These are the ones that are significant.
top_results <- treatment_results[1:102, ]
top.eset <- GSE120615.rma[row.names(exprs(GSE120615.rma)) %in% row.names(top_results)]
phenotype.colors <- unlist(lapply(GSE120615.rma$phenotype, 
                                  function (x) {if (x == "Control") "red" 
                                                else if (x == "TwoHours") "navy"
                                                else "green"}))
heatmap.2(exprs(top.eset), col = viridis(256, option = "B"),
          trace = "none", keysize = 1.5, key.title = NA,
          ColSideColors = phenotype.colors)
legend("topleft", legend = c("Control", "TwoHours", "SixteenHours"), fill = c("red", "navy", "green")) 

# In the heatmap, there are two main clusters: SixteenHours and Control/TwoHours.
# This shows that the results at 16 hours of virus exposure are distinct and significant from control and 2-hour treatments, which are clustered together. 
```

* Use GSEA to check which functional pathways are involved in smoking response. Use KEGG and REACTOME pathways from MSigDB as reference ontologies. Create barplots of NES for **independent** pathways with adjusted p-value < 0.05 and absolute NES > 2.
```{r, error=TRUE, warning=FALSE, message=FALSE}
library(fgsea)
library(data.table)
library(ggplot2)

# Read in pathway files in .gmt format using the function fgsea::gmtPathways
kegg <- gmtPathways("./c2.cp.kegg.v7.4.symbols.gmt")
reactome <- gmtPathways("./c2.cp.reactome.v7.4.symbols.gmt")
biocarta <- gmtPathways("c2.cp.biocarta.v7.4.symbols.gmt")
pid <- gmtPathways("c2.cp.pid.v7.4.symbols.gmt")
wikipathways <- gmtPathways("c2.cp.wikipathways.v7.4.symbols.gmt")
curated <- gmtPathways("c2.all.v7.4.symbols.gmt") # C2: curated gene sets, contains CGP and CP data sets
pathways <- c(kegg, reactome, biocarta, pid, wikipathways, curated)

# Read in the differential expression result file from above
DE_res <- read.table("./GSE120615_DE_results.txt", header = T, sep = "\t")

# Remove probes with no corresponding gene symbols (i.e. SYMBOL==“NA”)
# Retain statistics of the gene symbol with the smallest P.Value
library(dplyr)
DE_res <- DE_res %>%
  filter(!is.na(SYMBOL)) %>%
  group_by(SYMBOL) %>%
  arrange(P.Value) %>%
  filter(row_number() == 1) %>%
  data.frame()

# Rank the gene list based on their t-statistics
gene_list <- DE_res$t
names(gene_list) <- DE_res$SYMBOL
gene_list <- sort(gene_list, decreasing = T)

# Include pathways with >=3 and <=500 genes and weigh each gene based on its t-statistic (gseaParam = 1).
fgseaRes <- fgsea(pathways = pathways, stats = gene_list, minSize = 3, maxSize = 500, gseaParam = 1)
fgseaRes <- fgseaRes[order(fgseaRes$pval), ]

# Select independent pathways using fgsea::collapsePathways
# Obtain final GSEA output
collapsedPathways <- collapsePathways(fgseaRes = fgseaRes[pval < 0.05], pathways = pathways, stats = gene_list)
mainPathways <- fgseaRes[pathway %in% collapsedPathways$mainPathways] # keep results of independent pathways

# Convert pathways to factors where the levels are the same as pathway order
mainPathways$pathway <- factor(mainPathways$pathway, levels = mainPathways$pathway)
ggplot(mainPathways, aes(y = NES, x = pathway)) +
  geom_bar(width = 0.8, position = position_dodge(width = 0.8), stat = "identity", fill = "blue") + 
  coord_flip() +
  theme_bw() +
  theme(axis.title.y = element_blank())
```

* What is the top pathway enriched in genes with **increased expression** in response to smoking based on the **NES value**? Plot the leading edge and list the genes that drive the enrichment of this pathway. (*3 points*)
```{r, error=TRUE, warning=FALSE, message=FALSE}
# The top pathway enriched in genes with increased expression in response to smallpox virus is WP_COMPLEMENT_ACTIVATION.

# Plot leading edge for WP_COMPLEMENT_ACTIVATION pathway using fgsea::plotEnrichment
plotEnrichment(pathway = pathways[["WP_COMPLEMENT_ACTIVATION"]], stats = gene_list) + 
  labs(title="WikiPathways Complement Activation")

# List all leading edge genes for REACTOME_ARACHIDONIC_ACID_METABOLISM pathway
mainPathways %>% 
  filter(pathway=="WP_COMPLEMENT_ACTIVATION") %>% 
  dplyr::select(leadingEdge) %>% 
  unlist() %>% 
  unname()
```

### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.
```{r, error=TRUE, warning=FALSE, message=FALSE}

```
